### ISSUE TITLE
Validating optional fields

### ISSUE THREAD

**Issue by susickypavel** [fabian-hiller/modular-forms#5]
Hi 👋,

I couldn't find any info according to validating optional fields. For example, I have an email validation, but the user can decide not to add an email, but validation requires it.

I did a workaround by creating a custom validator, but I just wondered whether I couldn't find it or if you decided not to include it in the library.

PS: Thank you for this library; it's great!

**Comment by fabian-hiller** [comment]
Thanks for your feedback! I hadn't thought of this use case before. However, I think that this can be implemented very easily. I will think about it and keep you updated here. Until then, I recommend you to use the [`custom`](https://modularforms.dev/api/custom) function.

**Comment by susickypavel** [comment]
### Option 1: The optional prop

This could be useful if you wanted to add a `required: boolean` prop to the `field.props`, which would help to improve accessibility for screen readers by announcing the field as required. However, I'm not entirely sure whether there are cases of form inputs that don't accept the required attribute and then this prop wouldn't make sense.

### Option 2: Optional by default

This seemed intuitive and was my first thought about how `required` works. 

### Option 3: The optional option

This would clutter the code in large forms, so I'm not very fond of this idea.

### Option 4: Leave it to the user

This could potentially lead to duplicated code across the codebase, and I think implementing the optional behaviour will improve DX significantly.

TLDR: 1 or 2 is great IMO, it depends on whether the required prop makes sense.

**Comment by fabian-hiller** [comment]
I agree that option 2 is the most intuitive. It is also similar to native form validation. It is my favorite so far. It will then be important to explain in the docs, for example with `minLength`, that no error is thrown on an empty string or list and the function must be combined with `required` if necessary. I think I will introduce this feature in v0.9 in the coming days. I will keep you up to date here.

I like your thought about the accessibility of the options 1. I have had similar thoughts as well. Maybe I can find a way to add this information to the validation functions and provide the attributes based on that via the `props` object. You are right, I need to check the native validation attributes of all input elements in this case, as well as integration with existing component libraries.

**Comment by susickypavel** [comment]
I think that, at the very minimum, the slider input type is probably an issue here. Maybe it's better to leave this attribute to the consumer of the lib, but if you come up with a smart solution, I'm all for it!

**Comment by fabian-hiller** [comment]
v0.9 is now online and validation of optional fields is now possible as discussed here. Thanks for your feedback!

**Comment by iamriajul** [comment]
@fabian-hiller How can I validate optional fields with valibot, I did try the optional function for creating schema but that didn't behave as expected!

**Comment by fabian-hiller** [comment]
Can you share your schema? If a field contains no input, an empty string is typically returned.

**Comment by iamriajul** [comment]
Originally tried:
```tsx
export const RegisterSchema = object({
  email: optional(string([
    email('Please enter a valid email'),
  ])),
  name: string([
    minLength(1, 'Please enter your name.'),
    minLength(3, 'Please enter a valid name.'),
  ]),
});
```

Workaround:
```tsx
export const RegisterSchema = object({
  email: optional(string([
    custom(
      value => !value || !!value.toString().match(/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/),
      'Please enter a valid email address.'
    ),
  ])),
  name: string([
    minLength(1, 'Please enter your name.'),
    minLength(3, 'Please enter a valid name.'),
  ]),
});
```

**Comment by fabian-hiller** [comment]
As mentioned above, the `<input />` element returns an empty string if there is no input. So `optional` has no effect. There are several workarounds/solutions but we may need an explicit solution for this in the future.

```ts
// With union and empty string literal
export const RegisterSchema1 = object({
  email: union([literal(''), string([email('Please enter a valid email')])]),
  name: string([
    minLength(1, 'Please enter your name.'),
    minLength(3, 'Please enter a valid name.'),
  ]),
});

// With coerce and nullable (works also with optional)
export const RegisterSchema = object({
  email: coerce(
    nullable(string([email('Please enter a valid email')])),
    (input) => (input || null) as any // I will improve the types here
  ),
  name: string([
    minLength(1, 'Please enter your name.'),
    minLength(3, 'Please enter a valid name.'),
  ]),
});
```

**Comment by iamriajul** [comment]
I've settled with `custom`, and as per the first suggestion to use union, I'm trying to avoid union, union has some serious bugs I think. As you might have noticed I've posted that on another issue: https://github.com/fabian-hiller/valibot/issues/216#issuecomment-1773779047

**Comment by fabian-hiller** [comment]
Ok. `union` has no bugs. In your example, Modular Forms were the problem. But you're right, `union` is problematic with forms because we don't know what error message to display if every union option returns issues.
